### PR TITLE
[Proposal] New API for Jetpack Compose

### PR DESCRIPTION
**Pull request by burnoo**
I'm starting a discussion on `#koin-dev` - feel free to join  if you have any feedback or questions regarding this PR.

### Motivation
I’ve been working with Koin for Compose for a while now and have a proposal to improve how `KoinContext` management works in `@Composable` functions. I believe the configuration of the application should be done in the top level composable. This way it is not dependent on Koin initialization in the `Application` class, which allows for easy `KoinContext` replacement in `@Preview` composables and tests. It introduces a new API - `Koin` and `KoinScope` composables.

### Usage example (with scopes and @Preview):
```kotlin

object ScopeTypeA
object ScopeTypeB

data class A(val value: String)

val scopedModule = module {
    scope<ScopeTypeA> {
        scoped { A("scopeA") }
    }
    scope<ScopeTypeB> {
        scoped { A("scopeB") }
    }
}

@Preview
@Composable
fun Test() {
    Koin(appDeclaration = { modules(scopedModule) }) {
        Column {
            KoinScope(getScope = { createScope<ScopeTypeA>() }) {
                Text(get<A>().value)
            }
            KoinScope(getScope = { createScope<ScopeTypeB>() }) {
                Text(get<A>().value)
            }
        }
    }
}
```

### Pros:
- scopes works
- easy `KoinContext` replacement in tests
- easy `KoinContext` replacement in `@Preview` composables
- no need to use `Application` class (which may help with Compose Multiplatform support in the future)
- backward compatibility (using `GlobalContext` as default value)

### This PR includes:
- new API implementation
- instrumented tests
- necessary library updates
- fix of nullability issue in `ViewModelComposeExt.kt`

### Problems
- I couldn't update Compose to 1.0.2 because it uses newer version of Kotlin (1.5.21) than Koin
- to run the project I had to update `androidx_compose_gradle_version` to `7.0+`

### Issues fixed with this PR:
- https://github.com/InsertKoinIO/koin/issues/1177
- https://github.com/InsertKoinIO/koin/issues/1119
- https://github.com/InsertKoinIO/koin/issues/1072
- https://github.com/InsertKoinIO/koin/issues/823 (partially by `getKoinComponent().run { /* inject works here /* }`)

### TODO
- javadocs if proposal is accepted

### CODE REVIEW

**Comment by burnoo** [comment]
I've released this changes as standalone library for Compose Multiplatform. You can check it out here: https://github.com/burnoo/cokoin

**Comment by arnaudgiuliani** [comment]
Yep, really appreciate your contribution. Could we add an update to https://github.com/InsertKoinIO/compose-samples?
It's a fork of base sample projects. It would help illustrate how we can play with Koin context in @Composable functions.

I can provide some guidelines for the javadoc.

File: android-compose/koin-androidx-compose/src/main/java/org/koin/androidx/compose/ComposeExt.kt [hunk 1]
```diff
@@ -17,30 +17,33 @@ package org.koin.androidx.compose
...
- * @param qualifier
- * @param parameters
- *
- * @return Lazy instance of type T
- *
- * @author Arnaud Giuliani
- * @author Henrique Horbovyi
- */
 @Composable
-inline fun <reified T> get(
```
Comment (arnaudgiuliani) [inline android-compose/koin-androidx-compose/src/main/java/org/koin/androidx/compose/ComposeExt.kt:36]: Why replace the `get` function? Either we can deprecate it or keep it 🤔

**Comment by burnoo** [comment]
After opening this pull request, I thought that I want to do more with Koin for Compose. My first priority was to add Compose Multiplatform support. I also thought that `ViewModel` should be in the another artifact. I looked into issues related to Compose and found that Navigation is not supported.

With that in mind I've created my own library Cokoin - https://github.com/burnoo/cokoin. At first it was Multiplatform Compose, but then I've added Jetpack support. It contains the following features:
- this PR (Koin Composable + Scopes Support)
- Compose Multiplatform support
- separate artifact for `ViewModel`
- support for Navigation Compose (with separate artifact)

I think developing Koin for Compose as a separate library would have many benefits, such as:
- independence from the Koin version release cycle
- many artifacts would not overload Koin
- there would be no compatibility problems related to the Kotlin version
- easier keeping it up to date with the Jetpack Compose version
- Compose Multiplatform alpha support

I think that Cokoin could be official way of using Koin with Compose. I can do the following:
1. Add Javadocs to Cokoin
2. Create documentation page (docusaurus same as Koin)
3. Give you an opportunity to check the code (CR) and make adjustments if necessary
4. Add backward compatibility (for now in cokoin there is no backward compatibility, because `GlobalContext` is not in core, should be easy to add)
5. Update all Compose examples with Cokoin (and maybe move it to Cokoin repository)
6. (Maybe) Close this PR and deprecate `org.koin:koin-androidx-compose`

@arnaudgiuliani What do you think?

File: android-compose/koin-androidx-compose/src/main/java/org/koin/androidx/compose/ComposeExt.kt (hunk 1 above)
Comment (burnoo) [reply android-compose/koin-androidx-compose/src/main/java/org/koin/androidx/compose/ComposeExt.kt:36]: I only changed its place in code and removed the docs (by accident). It's here https://github.com/InsertKoinIO/koin/pull/1178/files/6062bdfdb3b2fc25f38acb26a0b8995b9480e508#diff-f9db5f6b8a1cbf8a2b9601a97856b228a48df598a9fb60c8e451a43b8ce5ccd2R41

**Comment by ZeDzislaw** [comment]
On one side im happy that someone picked up compose koin DI problems, but on other side i'm worried that actual split from official repo will end up dead. Wouldn't be better to make effort to prepare PR to merge it to this repository? I understand some of your points for separate lib, but again - Im afraid of comunity dispersion.

**Comment by burnoo** [comment]
@ZeDzislaw
Unfortunately this repository is not actively developed lately. Last Koin version was released 3 months ago. Compose is a very new technology and I think koin-compose releases should be much more frequent.

In my library I Introduced 5 more artifacts (and I am considering adding more for `SavedState` support). Adding them to this repository might not be the best idea.

I think a good compromise would be to move `cokoin` to InsertKoinIO GitHub, change artifacts name to `io.insert-koin:cokoin-*` (or other name) and put me in charge of managing releases. By doing so:
- I colud release the library frequently
- I could introduce a few artifacts and do not overload Koin
- cokoin would be a part of official GitHub, so would have community support

**Comment by ZeDzislaw** [comment]
@burnoo I agree. I myself was concerned about the lack of new commits, especially in the context of jetpack compose. I appreciate your work on this subject and we will probably give some feedback in a couple of weeks as we plan to base the new screens on jetpack compose navigation in conjunction with koin.

**Comment by apkelly** [comment]
Is the plan to move all of my Koin initialisation to the main Composeable in my Android app, as I don't think this covers some of the use-cases we need. Initialising my Koin modules in the Application class means I can inject my implementations into components that aren't Composables e.g. Services

**Comment by burnoo** [comment]
@arnaudgiuliani Have you read my previous comments? What do you think?

@apkelly You are right. To use Koin in Services you have to initialize it in Application class. Sharing code between Service and Composables can be done by using same modules in Application and Koin Composable:
```kotlin
// Application class
override fun onCreate() {
    super.onCreate()
    startKoin { modules(sharedModule, servicesModule) }
}
```
```kotlin
@Composable
fun App() {
    Koin(appDeclaration = { modules(sharedModule, uiModule) }) {
       //...
    }
}
```

**Comment by arnaudgiuliani** [comment]
> @arnaudgiuliani Have you read my previous comments? What do you think?

we can perhaps consider decoupling compose related development to help go faster. I'm open to help integrate in a better way

**Comment by arnaudgiuliani** [comment]
Closing `master` branch closed all PRs :/ really sorry. Can you update your PR against `main` branch

**Pull request closed without merging**